OpenSim/Examples/DataTable/example.cpp

```diff
@@ -0,0 +1,357 @@
+ * Copyright (c) 2005-2015 Stanford University and the Authors                *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0.         *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include <OpenSim/Common/DataTable.h>
+
+// Function to do nothing more than  suppress 'un-used variable' warning. 
+// Ignore these calls.
+template<typename AnyType>
+void ignore(AnyType&) {}
+
+
+// This file demonstrates the use of the OpenSim::DataTable_ class. 
+// See OpenSim::DataTable_ for details on the interface. 
+// OpenSim::DataTable_ is a container of a matrix of elements(of any type)
+// that supports random-access to the columns and rows through the their
+// index. Columns can be labeled with std::string and accessing the columns
+// through labels is constant-time on average. DataTable_ can also hold key-
+// value pairs as metadata where key is a std::string and value can be of any
+// type.
+
+int main() {
+  constexpr double EPSILON{0.00001};
+
+  // Default construct a DataTable of SimTK::Real[alias for double].
+  OpenSim::DataTable_<SimTK::Real> dt_real{};
+  // Default construct a DataTable of SimTK::Vec3[a tuple of 3 double(s)].
+  OpenSim::DataTable_<SimTK::Vec3> dt_vec3{};
+  // Default construct a DataTable of SimTK::Vec6[a tuple of 6 double(s)].
+  OpenSim::DataTable_<SimTK::Vec6> dt_vec6{};
+
+  std::vector<SimTK::Real> data_real{};
+  std::vector<SimTK::Vec3> data_vec3{};
+  std::vector<SimTK::Vec6> data_vec6{};
+  for(int i = 0; i < 12; ++i) {
+    data_real.push_back(i);
+    data_vec3.push_back(SimTK::Vec3{1, 2, 3} + i);
+    data_vec6.push_back(SimTK::Vec6{1, 2, 3, 4, 5, 6} + i);
+  }
+
+  // Construct a DataTable using iterators. We can arrange the elements
+  // row-wise or column-wise using the argument 'dir'. The argument 'ndir' is
+  // required in either case to let the DataTable know about the shape of the
+  // matrix. The following creates a 4x3 DataTable filling the elements
+  // column-wise.
+  {
+    OpenSim::DataTable_<SimTK::Real> real{data_real.cbegin(),
+                                          data_real.cend(),
+                                          3,
+                                          OpenSim::COLWISE};
```

Comment:
Typed constants should not use `ALLCAPS`, that ugly convention should be reserved for preprocessor symbols. The reason for it is to call attention to the different rules that apply to preprocessing symbols than the strongly-typed in-language features. This should be `OpenSim::ColWise` instead. I'm not sure we have a discussion about that in OpenSim's coding standards but Simbody has one [here](https://github.com/simbody/simbody/blob/master/CONTRIBUTING.md#constants).